_posts/2016-06-27-bits-of-shapeless-2.md

```diff
@@ -0,0 +1,519 @@
+```
+
+Testing it:
+
+```scala
+Person("John", 30).toJson
+// res35: jawn.ast.JValue = {"age":30,"name":"John"}
+
+case class Address(Street: String, Number: Option[Int])
+// defined class Address
+
+Address("Small Street", None).toJson
+// res36: jawn.ast.JValue = {"Number":null,"Street":"Small Street"}
+
+Address("Big Street", Some(22))
+// res37: Address = Address(Big Street,Some(22))
+```
+
+### Deriving sealed hierarchies
+
+The canonicall way to represent **ADTs** -- abstract data types -- in scala is using sealed traits and case classes/objects, such as the following:
+
+```scala
+sealed trait Shape extends Product with Serializable
+case class Square(side: Long) extends Shape
+case class Rectangle(width: Long, eight: Long) extends Shape
+case class Circle(radius: Long) extends Shape
+```
+
+We can render a square, circle or rectangle to json, since they are case classes and we already know how to render those:
+
+```scala
+Circle(2).toJson
+// res38: jawn.ast.JValue = {"radius":2}
+
+Square(4).toJson
+// res39: jawn.ast.JValue = {"side":4}
+
+Rectangle(2,4).toJson
+// res40: jawn.ast.JValue = {"eight":4,"width":2}
+```
+
+However, if we don't know what kind of shape we have there is a problem:
+
+```scala
+def shapeToJson(s: Shape) = implicitly[JsonWrites[Shape]].write(s)
+// <console>:36: error: could not find implicit value for parameter e: JsonWrites[Shape]
+//        def shapeToJson(s: Shape) = implicitly[JsonWrites[Shape]].write(s)
+//                                              ^
+```
+
+To solve this problem there is the concept of a **Coproduct**, the dual concept of a product.
+As oposed to the product, a coproduct represents one element that belong to just one of a set of types. It is like scale's Either extended to an arbitrary number of alternatives. In Shapeless:
+
+```scala
+type stringOrIntCP = String :+: Int :+: CNil
+// defined type alias stringOrIntCP
+```
+
+This type represents something that can be a string or an integer.
+
+Let us see what the coproduct definition looks like in shapeless:
+
+```scala
+sealed trait Coproduct extends Product with Serializable
+
+sealed trait :+:[+H, +T <: Coproduct] extends Coproduct {
+  // Non-recursive fold (like Either#fold)
+  def eliminate[A](l: H => A, r: T => A): A
+}
+
+final case class Inl[+H, +T <: Coproduct](head : H) extends :+:[H, T] {
+  override def eliminate[A](l: H => A, r: T => A) = l(head)
+}
+
+final case class Inr[+H, +T <: Coproduct](tail : T) extends :+:[H, T] {
+  override def eliminate[A](l: H => A, r: T => A) = r(tail)
+}
+
+sealed trait CNil extends Coproduct {
+  def impossible: Nothing
+}
+```
+
+It's a bit more involved than `HList`.
+
+* As analogous to HList's `HNil`, `CNil` marks the end of the type definition, and is used to stop recursion and to enable proper type construction.
+*`:+:` is analogous to Hlist's '::` type constructor, however it can be either an `Inl`or `Inr`.
```

Comment:
add space after * to properly render the bullet list. also the :: is not properly escaped